Tests/AsyncQueueTests/AsyncQueueTests.swift

```diff
@@ -26,8 +26,202 @@ import XCTest
+                await counter.incrementAndExpectCount(equals: 2)
+            }
+            await counter.incrementAndExpectCount(equals: 1)
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: 3)
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_async_retainsReceiverUntilFlushed() async {
+        var systemUnderTest: AsyncQueue? = AsyncQueue()
+        let counter = Counter()
+        let expectation = self.expectation(description: #function)
+        let semaphore = Semaphore()
+        systemUnderTest?.async {
+            // Make the queue wait.
+            await semaphore.wait()
+            await counter.incrementAndExpectCount(equals: 1)
+        }
+        systemUnderTest?.async {
+            // This async task should not execute until the semaphore is released.
+            await counter.incrementAndExpectCount(equals: 2)
+            expectation.fulfill()
+        }
+        // Nil out our reference to the queue to show that the enqueued tasks will still complete
+        systemUnderTest = nil
+        // Signal the semaphore to unlock the remaining enqueued tasks.
+        await semaphore.signal()
+
+        await waitForExpectations(timeout: 1.0)
+    }
+
+    func test_async_doesNotRetainTaskAfterExecution() async {
+        final class Reference: Sendable {}
+        final class ReferenceHolder: @unchecked Sendable {
+            var reference: Reference? = Reference()
+        }
+        let referenceHolder = ReferenceHolder()
+        weak var weakReference = referenceHolder.reference
+        let asyncSemaphore = Semaphore()
+        let syncSemaphore = Semaphore()
+        systemUnderTest.async { [reference = referenceHolder.reference] in
+            // Now that we've started the task and captured the reference, release the synchronous code.
+            await syncSemaphore.signal()
+            // Wait for the synchronous setup to complete and the reference to be nil'd out.
+            await asyncSemaphore.wait()
+            // Retain the unsafe counter until the task is completed.
+            _ = reference
+        }
+        // Wait for the asynchronous task to start.
+        await syncSemaphore.wait()
+        referenceHolder.reference = nil
+        XCTAssertNotNil(weakReference)
+        // Allow the enqueued task to complete.
+        await asyncSemaphore.signal()
+        // Make sure the task has completed.
+        await systemUnderTest.await { /* Drain the queue */ }
+        XCTAssertNil(weakReference)
+    }
+
+    func test_await_sendsEventsInOrder() async {
+        let counter = Counter()
+        for iteration in 1...1_000 {
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: iteration)
+            }
+
+            guard iteration % 25 == 0 else {
+                // Keep sending async events to the queue.
+                continue
+            }
```

Comment:
I don't follow why this code exists in the test. Can you help me understand?

Reply:
I wanted some way to interleave multiple `async` events with each `await` event that was deterministic. We could `await` after every `async`, but because the `await` method requires waiting for the work to finish I felt like the test wouldn't necessarily catch an issue where `await` wouldn't wait for every enqueued event but would just wait for the first enqueued event...

Short version: this `guard` could be removed and it'd probably be fine. But I kinda like emitting multiple `async` events for every `await` check.